dna/zomes/courses/code/src/course.rs

```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
 pub fn create(title: String, timestamp: u64) -> ZomeApiResult<Address> {
-    let anchor_entry = anchor_entry();
-    let anchor_address = hdk::commit_entry(&anchor_entry)?; // if Anchor exist, it returns the commited one.
-
-    let new_course = Course::new(title, AGENT_ADDRESS.to_string().into(), timestamp);
-    let new_course_entry = new_course.entry();
-    let new_course_address = hdk::commit_entry(&new_course_entry)?;
-
-    hdk::link_entries(&AGENT_ADDRESS, &new_course_address, "teacher->courses", "")?;
-
-    hdk::link_entries(&anchor_address, &new_course_address, "course_list", "")?;
+    // create anchor for all courses
+    // it's an idempotent action that won't change anything if this anchor already exists in DHT
+    let anchor_entry = anchor_all_courses_entry();
+    // if this entry already exists, we'll get it's address without creating anything new
+    let anchor_address = hdk::commit_entry(&anchor_entry)?;
+
+    // create a CourseAnchor entry and commit it to DHT so we have it's address
+    let course_anchor_entry =
+        CourseAnchor::new(AGENT_ADDRESS.to_string().into(), timestamp).entry();
+    let course_anchor_address = hdk::commit_entry(&course_anchor_entry)?;
+
+    // link CourseAnchor to anchor_all_courses for this course to be discoverable
+    hdk::link_entries(&anchor_address, &course_anchor_address, "course_list", "")?;
+
+    // create a CourseData entry & commit it to DHT
+    let course_data_entry = CourseData::new(
+        title,
+        AGENT_ADDRESS.to_string().into(),
+        course_anchor_address.clone(),
+        timestamp,
+    )
+    .entry();
+    let course_data_address = hdk::commit_entry(&course_data_entry)?;
+
+    // link CourseAnchor to the CourseData entry for it to be discoverable
+    hdk::link_entries(
+        &course_anchor_address,
+        &course_data_address,
+        "course_anchor->course_data",
+        "",
+    )?;
 
-    Ok(new_course_address)
+    Ok(course_anchor_address)
 }
 
 pub fn update(
     title: String,
     modules_addresses: Vec<Address>,
     course_address: &Address,
 ) -> ZomeApiResult<Address> {
-    let course: Course = hdk::utils::get_as_type(course_address.clone())?;
+    //TODO:  check if addr we receive is actually CourseAnchor, not CourseData
 
-    let new_version_course = Course::from(
+    // get latest CourseData entry by looking at links of CourseAnchor entry
+    let latest_course_data_addr =
+        get_latest_link_addr(course_address, "course_anchor->course_data")?;
+    let course_data: CourseData = hdk::utils::get_as_type(latest_course_data_addr)?;
+
+    // create a new version of CourseData entry
+    let new_course_data_entry = CourseData::from(
         title,
-        course.teacher_address,
-        course.timestamp,
+        course_data.teacher_address,
+        course_data.course_anchor,
+        // TODO: we shouldn't be using old timestamp here. Maybe we should receive it as a parameter for update?
```

Comment:
Yes! Definetly, we should receive timestamp as a parameter to avoid update cycles.